Makefile

```diff
@@ -13,13 +13,25 @@ override BRANCH_ROOT += $(addsuffix /branch,$(MODULE_ROOT))
 override BTB_ROOT += $(addsuffix /btb,$(MODULE_ROOT))
 override PREFETCH_ROOT += $(addsuffix /prefetcher,$(MODULE_ROOT))
 override REPLACEMENT_ROOT += $(addsuffix /replacement,$(MODULE_ROOT))
+#allows for swappable dram controller (in a future commit, this should probably become a true module)
+override DRAM_CONTROLLER_ROOT += $(addsuffix /$(DRAM_MODEL),$(addsuffix /dram_controller,$(MODULE_ROOT))) 
```

Comment:
First of all, I'm _very_ dubious that the controller ought to be a true module. The existing modules are all things that the user might write themselves and swap in/out for experiments. This is not that. People have disagreed with me on this point, though, particularly when we were talking about inclusive/exclusive caches. I'm willing to be convinced here, but the mechanisms for modularity are so clunky that I'm hesitant to add more of them.

Also, I don't think I like the design of selectively compiling based on path name here. In my experience, we should compile **all** of the code **every** time. Otherwise, code will languish and break from disuse over time. See the work that I did converting the debug prints from `#ifdef` (selectively compiling source code) to `if constexpr` (compiled, then removed). If we build in selective compilation here, it will lead to maintenance headaches in the future.

Reply:
When I decided to try to integrate Ramulator, I knew in my mind that I shouldn't tie ChampSim's compilation to Ramulator. We take all the baggage of Ramulator's compilation and add it to ChampSim.
Note that the workflows I am running for Ramulator, only a few of the compilers are run. If there isn't a way to separate them, we can't compile with those excluded compiler versions either.

If we are talking in terms of a piece of code we compile in but selectively use, then that sounds a lot like how the module system currently works to me.

Reply:
I think the main benefit from making the dram controller a module is that we could lay the groundwork for introducing a network on chip interface, but that's probably best left for a different patch and in-depth discussion.

Reply:
I agree that there's potential for a future direction, but I think we should keep our changes minimal until we can clarify the requirements for such a change. See my recent thinking [here](https://github.com/ChampSim/ChampSim/issues/424#issuecomment-2353592540).